jenkins/Jenkinsfile

```diff
@@ -0,0 +1,228 @@
+def install_test_dependencies(String cache_prefix) {
+  /* restore cache with key: << parameters.cache_prefix >>-dependencies-{{ checksum "dev-requirements.txt" }}-{{ checksum "test-requirements.txt" }}-{{ checksum "setup.py" }} */
+  sh '''#!/bin/bash
+  pip install -r dev-requirements.txt --user
+  pip install -r test-requirements.txt --user
+  pip install -e '.[dispatcher]' --user
+  '''
+  /* save cache with relevant paths and key */
+}
+def pytest(String target){
+  echo "running pytest << ${target} >>"
+  sh """#!/bin/bash
+  pytest \
+    -n auto \
+    --cov-report term-missing \
+    --cov= ${target} \
+    ${target}/tests \
+    --junitxml=test-results/ ${target}
+  """
+}
+def build_docs(String project){
+  echo "build docs for ${project}"
+  sh """
+  #!/bin/bash
+  sphinx-build \
+    -d _build/doctrees \
+    -D version="\$(python setup.py --version)" \
+    docs/${project} \
+    docs/${project}/_build/html
+  """
+}
+def upload_docs(String project){
+  echo "upload docs to S3 for ${project}"
+  uploadToReleaseS3(
+    "${env.WORKSPACE}/project-docs/docs/${project}", 
+    "${env.VERSION}", 
+    "${env.PRERELEASE}", 
+    "${env.UPLOAD_FOLDER}", 
+    BRANCH_BUILD_NUMBER, 
+    "${env.BUILD_NUMBER}"
+  )
+}
+@Library('pipeline-shared-library') _
+pipeline {
+  parameters {
```

Comment:
why you need those params? you can get them from branch:

environment {
    VERSION = getVersion("${env.BRANCH_NAME}").trim()
    PRERELEASE = getPreRelease("${env.BRANCH_NAME}").trim()
  }